api/src/main/java/marquez/db/DbRetention.java

```diff
@@ -0,0 +1,112 @@
+
+  /* Tables to apply retention policy on (with cascade delete constraint). */
+  private static final Set<String> TABLES = Set.of(DATASETS, JOBS, LINEAGE_EVENTS);
+
+  public static void retentionOnDbOrError(@NonNull Jdbi jdbi, final int retentionDays)
+      throws DbRetentionException {
+    for (final String table : TABLES) {
+      log.info("Applying retention policy of '{}' days to '{}' table...", retentionDays, table);
+      // Build, then execute retention policy.
+      final int totalNumOfRowsDeleted =
+          executeDbRetentionQuery(
+              jdbi,
+              buildQuery(
+                  """
+                  DELETE FROM %s
+                    WHERE %s IN (
+                      SELECT %s FROM %s
+                       WHERE %s < CURRENT_TIMESTAMP - INTERVAL '%d days'
+                       LIMIT %d
+                   ) RETURNING %s;
+                  """,
+                  table, retentionDays));
+      // Determine whether any rows met the retention policy.
+      if (totalNumOfRowsDeleted == 0) {
+        log.info("No rows older than '{}' days in '{}' table.", retentionDays, table);
+        continue;
+      }
+      log.info("Successfully deleted '{}' rows from '{}' table.", totalNumOfRowsDeleted, table);
+    }
+  }
+
+  /** Returns the number of {@code row}s deleted by the provided retention {@code query}. */
+  private static int executeDbRetentionQuery(
+      @NonNull Jdbi jdbi, @NonNull final String retentionQuery) throws DbRetentionException {
+    try {
+      return jdbi.withHandle(
+          handle -> {
+            final Stopwatch stopwatch = Stopwatch.createStarted();
+            int totalNumOfRowsDeleted = 0; // Keep count of rows deleted.
+            boolean hasMoreRows = true;
+            while (hasMoreRows) {
+              // Apply batch deletes (for improved performance) on rows in table with a creation
+              // time > date allowed in retention query.
+              final List<UUID> rowsDeleted =
+                  handle.createQuery(retentionQuery).mapTo(UUID.class).stream().toList();
+              final int numOfRowsDeleted = rowsDeleted.size();
+              totalNumOfRowsDeleted += numOfRowsDeleted;
+              hasMoreRows = !rowsDeleted.isEmpty();
+            }
+            // Measure the time elapsed to execute retention query, then return number of rows
+            // deleted.
+            stopwatch.stop();
+            log.debug(
+                "Deleted '{}' rows in '{}' secs using retention policy: {}",
+                totalNumOfRowsDeleted,
+                stopwatch.elapsed().toSeconds(),
+                retentionQuery);
+            return totalNumOfRowsDeleted;
+          });
+    } catch (StatementException errorOnDbRetentionQuery) {
+      log.error("Failed to apply retention query: {}", retentionQuery);
+      // Propagate throwable up the stack.
+      throw new DbRetentionException(errorOnDbRetentionQuery.getCause());
```

Comment:
if you log before rethrowing you'll have double logs.
Also I don't understand why you don't chain the exception itself instead of the cause.
I think you should just throw here and add the query to the new message.
Something like:
`throw new DbRetentionException("Failed to apply retention query: " + retentionQuery errorOnDbRetentionQuery);`

Reply:
Ok yeah, I agree that this will double log and make things harder to debug by polluting the logs with, essentially, the same error.